pkg/operator/operator.go

```diff
@@ -632,16 +636,38 @@ func (o *Operator) sync(ctx context.Context, key string) error {
 		klog.Errorf("error occurred while setting status to in progress: %v", err)
 	}
 
-	taskErrors := tl.RunAll(ctx)
```

Comment:
because there is a run - step and a validation step

Reply:
I see. What is the motivation for separating these two steps? The current "validation" is part of the task run, wouldn't that work for the more extensive validation as well?

The reason I'm asking is that this does add additional code that is close to a duplication of existing code (error handling after `RunAll` and `ValidateAll`, and more concerning imho `ValidateAll` and `ValidateConcurrently` in the `task` module).
I would also argue that linking resource creation and subsequent validation make sense in the same task, since one should happen after the other. By running them independently we have to handle an additional error case (https://github.com/openshift/cluster-monitoring-operator/pull/1558/files#diff-b0ca3a050a076f5a4007692cff8088723c9fe6f8c6734f8a8bc1bdd54f8d624cR423).
I feel like I'm missing something.

Reply:
> The current "validation" is part of the task run, wouldn't that work for the more extensive validation as well?

That is the initial thought I had as well, but when Simon and I discussed, we felt this approach is the better one to take since it separates the concerns correctly. What we concluded was that there are 2 concerns here 

1. Create / update resources: These are non blocking and independent steps which are fire and forget and should be non-blocking. E.g. the task that applies the prometheus resources shouldn't need to validate (blocking `wait.Poll()` ). Its responsibility is to ensure that the resources can be created and updated properly, failing which means the operator is in a `degraded` state.

2. `Validate` if the system is in the right state and ideally these should can run independent of the `Run` albeit in this PR I do cache prometheus in `t.prometheus`


> The reason I'm asking is that this does add additional code that is close to a duplication of existing code (error handling after `RunAll` and `ValidateAll`, and more concerning imho `ValidateAll` and `ValidateConcurrently` in the `task` module).

I do think we can keep the code dry at the cost of increased complexity and since it's an internal implementation, we can change it anytime.

> I would also argue that linking resource creation and subsequent validation make sense in the same task, since one should happen after the other.

IMHO, validation of a system's state shouldn't require a create/update step to have been executed. I feel these are independent of one another.

Reply:
I have strong feelings about being able to maintain things in the future. Most other concerns should serve that in my opinion :)

Talking about an individual task (say `PrometheusOperator`) I agree that Create/Update has no dependency on a Validate step. I would however argue that Validate entirely depends on the state we want and the state we have. Validate needs an input to check the actual state against. In CMO this expected state is very predictable and static (we don't expect any other actors to make any changes), it is in fact the input to Create/Update.
As you say yourself:
> Validate if the system is in the right state and ideally these should can run independent of the Run albeit in this PR I do cache prometheus in t.prometheus

Validate needs an input to compare the actual state of the world against.

To accomplish what the PR title says we can clearly distinguish between errors in the Create/Update step and validation errors. This can probably be done with custom errors alone. If we have dependencies between tasks, we should run these tasks in different task groups (as we are doing already).

And just a pedantic point: `Run` is just named `Run` as part of the task runner implementation. What that task runs is what we are discussing. Imho a task should contain the Create and Validate steps. If there are good arguments for distinguishing these steps and running them independently, we should have create tasks and validation tasks (one for every task we currently have). I don't think the task runner needs to know what a validation is, that knowledge belongs into the validation task.

Reply:
> I have strong feelings about being able to maintain things in the future. Most other concerns should serve that in my opinion :)

Well, the feeling is mutual, but IMHO, the premise feels faulty. It already starts with an assumption  that the code is unmaintainable rather than providing arguments that leads the me to understand why it is so  and leaves me to defend why it is more maintainable  😂

IOW, it fails to make the argument that making the `validation` separate from create/update leads to an unmaintainable state while providing reasons why keeping them close leads to more maintainable code. 

> Talking about an individual task (say PrometheusOperator) I agree that Create/Update has no dependency on a Validate step. 

Thinking about it, there is a actually a dependency that the validation "needs to know" the state expected by the `run` step; i.e. it should have the knowledge of what `run` step intended the system's state to be for it to validate **correctly** and therefore these aren't strictly independent. However this can be rectified by allowing the operator that makes both calls (Run, Validate) to consolidate the errors. 


### Scenario

Given: there already exists a `prometheus` CR and its pods are running fine.

#### Using the current implementation 

1. Now CMO runs  `prometheus` Task ( increase replica count for instance) 
2. the `run` step fails to apply `promethues` CR (for whatever reason). 
3. It sets the state as `degraded` 
4. Operator waits for all task `run`s to finish 
5. Operator runs `validate` step
6. The validate step gets the existing (but old) `prometheus` CR
7. It checks if the pods are running fine (which they are) 
8. Validates reports that system is `{ Degraded: false, Available: true }`.

This contradicts what the `run` step intents the system's state to be. 

In the current implementation, since the `run` step returned `degraded` and the `validation` step returned `fine`, CMO would report `{Degraded: true, Available true}`

NOTE: There is  inconsistency between what `run` step reports and what `validate` reports but it is consolidated by the `operator` and reports the state correctly.

Also note that the current structure ensures that all Tasks are validated and isn't the responsibility of an individual task to ensure that validation is performed in case of a failure to create/update. 

Now let's take the implementation that is being suggested and compare it to the approach in this PR. 

#### Combine Create and Validation into `Run`

1. CMO runs  `prometheus` Task ( increase replica count for instance) 
2. the `run` step failed to apply `promethues` CR (for whatever reason). 
4. It returns the state as `degraded`. ( see unless... )
5. The validate step won't even run, so the state isn't validated 

CMO would report `{Degraded: true, Available: ???}`

#### Unless every `Run` Task is structured differently

```go
type TaskError struct {
  Unknown bool
  Unavailable bool
  Degraded bool

}

func (t Task) Run() TaskError {
   var creationError = t.create()
   var validationError = t.validate()
   return combine(creationError, validationError)
}
```

Issues: 
Here the task instead of the `operator` ensures that the validation is run after create and combines the error but the knowledge that the `validation` **has to be run after `create`** is lost. 

Advantages: 
* Can Piggy back on current implementation of `Task.Run` albeit all `Run` need change to return the `TaskError` 

I am happy to take this approach but I will wait for a consensus before reorganising the code. 

cc: @jan--f @simonpasquier

Reply:
Thank you for elaborating! But me apologize first.

> IOW, it fails to make the argument that making the validation separate from create/update leads to an unmaintainable state while providing reasons why keeping them close leads to more maintainable code.

By no means did I intent to say this code is unmaintainable, only that I think we can improve it and make the life of future developer easier. For example duplicate code will likely drift in future iterations and thus increase complexity. A `Task`, that so far only knows how to `Run` something, now learns about an individual tasks internals (that it has a validation). This increases the context a developer has to keep when thinking about tasks, a step that imho should not be taken lightly.

Your `Task` restructuring proposal makes perfect sense to me. This keeps with how the tasks are structured today but with better error reporting and improved validation, as your PR proposes. I'm not sure I follow the issue you point out. The knowledge that validation has to happen after create is not lost, its encoded in the task at hand. I think this is the right place for it too.
The operator however encodes how task groups depend on each other. When two tasks depend on each other they should be run in two subsequent task groups.
The current implementation blurs this a bit though. By running _all_ create tasks before validation we would introduce some noise into the status reporting, e.g. creating a Prometheus CR will have no effect if creating the initial prometheus-operator deployment failed, yet both would be reported.

Reply:
Hopefully, this change capture what we discussed 🤞